### PR TITLE
Update NAT config page to address https://github.com/ipfs/ipfs-docs/issues/1424

### PR DESCRIPTION
**Pull request by ElPaisano**
# [Latest draft link](https://bafybeieoqz7uzhcghfu56wr4p7z53hvyutn3zvkxhoflyz5saqzgacfr24.on.fleek.co//how-to/nat-configuration/)

# Describe your changes
Addresses https://github.com/ipfs/ipfs-docs/issues/1424

# Files changed
docs/how-to/nat-configuration.md

# What issue(s) does this address?
 https://github.com/ipfs/ipfs-docs/issues/1424

# Does this update depend on any other PRs?
No

## Checklist before requesting a review
- [x] Passing the beta version of the **Check Markdown links for modified files** check. Action results can be viewed [here](https://github.com/ipfs/ipfs-docs/actions/workflows/action.yml).

## Checklist before merging
- [x] Passing all required checks (The beta **Check Markdown links for modified files** check is not required)

### CODE REVIEW

File: docs/how-to/nat-configuration.md [hunk 1]
```diff
@@ -1,32 +1,136 @@
...
+Is there a recommended repo or libp2p docs link that we can link to here that explains how to set this up for Kubo? 
+
+https://github.com/libp2p/specs/blob/master/relay/DCUtR.md doesn't seem right
+
+https://docs.libp2p.io/concepts/nat/hole-punching/#phase-ii-hole-punching? seems to be mostly conceptual
+
+https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/ seems to just explain the theory
+
+Maybe https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html and tell the reader that they can follow this tutorial to set up DCUtR Holepunching using Rust? Not specific to Kubo but seems like its the best we've got right now?
+!-->
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:52]: Dcutr is enabled by default, the user does not have to do anything.
I think we just need to explain what it is the draw backs and / or link to the blog post.

Some people have hole punchable nats so they see some success rate, however they would have better connectivity with a port forward.

File: docs/how-to/nat-configuration.md [hunk 2]
```diff
@@ -1,32 +1,136 @@
...
+1. [Update the Kubo Configuration](#update-the-kubo-configuration).
+1. [Restart your Kubo node](#restart-your-kubo-node). 
 
-Each router has different options and solutions for port forwarding. Most router manufacturers have guides for setting up port forwarding on their devices. In general, the steps are something like this:
+First, open a port from the internet to your internal Kubo node.
+
+#### Open a port 
+
+:::tip
+Assuming you're not trying to expose your daemon's API to the public internet, opening port `4001`/`tcp` should be sufficient.
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:68]: As a user, this sentence lead me to belive exposing the api port could be something i could want to do, which no, never, dont do that.

File: docs/how-to/nat-configuration.md [hunk 3]
```diff
@@ -1,32 +1,136 @@
...
+:::
+
+Each router has different options and solutions for port forwarding. Most router manufacturers have guides for setting up port forwarding on their devices. In general, the steps are:
 
 1. Log into your router.
 1. Locate your routers port forwarding section.
 1. Enter the IP address of your IPFS node.
-1. Set traffic to go through port `4001`.
+1. Set traffic to go through an outside port. `4001` is recommended if you are unsure.
 1. Reboot your IPFS node for the changes to take effect. Make sure to reboot the entire machine, not just the IPFS daemon.
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:77]: This is required later, after editing the config.

**Comment by ElPaisano** [comment]
@Jorropo I think I addressed your feedback, please take a look at the latest https://bafybeidbxtj5a5t6ashyyaq2j23gy6q2muga2yaftbkx2xxlgjkebmax6u.on.fleek.co/how-to/nat-configuration/ and let me know

File: docs/how-to/nat-configuration.md (hunk 1 above)
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md]: Alright, based the info off of kubo/docs and the blog post 
https://bafybeidbxtj5a5t6ashyyaq2j23gy6q2muga2yaftbkx2xxlgjkebmax6u.on.fleek.co/how-to/nat-configuration/#disable-dcutr-holepunching

File: docs/how-to/nat-configuration.md [hunk 4]
```diff
@@ -1,32 +1,170 @@
...
-## IPv6
+The appropriate configuration option for your router depends on your specific setup:
+
+- If your router supports them, [enable IPv6](#enable-ipv6) or [enable UPnP](#enable-upnp) to solve most connection issues
+- [Disable DCUtR holepunching](#disable-dcutr-holepunching), which is enabled by default as of Kubo v0.13
+- If IPv6 or UPnP are not available, [enable manual port forwarding](#enable-manual-port-forwarding)
+
+### Enable IPv6
 
 If your router and internet service provider (ISP) support IPv6, enabling it will mitigate some connection issues. Most modern routers have a single option to enable IPv6. We are unable to give detailed information about each router's settings and preferences here. Search your router manufacturer's website for _IPv6_ for information on how to enable it.
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:33]: Add a disclaimer here IPv6 does not always solve everything, my isp have per machine ipv6 but it still run a firewall that blocks inbound connections.
I belive that is fairly commun.

File: docs/how-to/nat-configuration.md [hunk 5]
```diff
@@ -1,32 +1,170 @@
...
 
-## UPnP
+### Enable UPnP
 
 If your router supports UPnP, IPFS will attempt to automatically allow inbound traffic to access your local content. Some home routers may need to be configured to explicitly enable UPnP. We are unable to give detailed information about each router's settings and preferences here. Search your router manufacturer's website for _UPnP_.
 
-## Port forwarding
 
-Each router has different options and solutions for port forwarding. Most router manufacturers have guides for setting up port forwarding on their devices. In general, the steps are something like this:
+### Disable DCUtR Holepunching
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:40]: You don't need to disable it, Kubo only turn it off if you are not reachable.
Doing a port forward will "automatically" disable this.

You don't need to disable it.
What I hope from this point is to explain that:
1. DCUtR have a 3~6s extra latency when connecting to a peer.
2. DCUtR does not always work.

And thus you probably want to seek a better alternative if you care about hosting content behind a NAT.

File: docs/how-to/nat-configuration.md [hunk 6]
```diff
@@ -1,32 +1,170 @@
...
 
-Each router has different options and solutions for port forwarding. Most router manufacturers have guides for setting up port forwarding on their devices. In general, the steps are something like this:
+### Disable DCUtR Holepunching 
+
+As of Kubo v0.13, [DCUtR hole punching is enabled by default](https://github.com/ipfs/kubo/blob/master/docs/changelogs/v0.13.md#-relay-v2-client-with-auto-discovery-swarmrelayclient).
+
+In the current version of DCUtR holepunching:
+
+- Performance is normal when a direct connection is established.
+- Performance is sub-optimal when someone connects to you, because all connections must go through a relay first, which adds an average delay of 5 seconds.
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:47]: I'm not sure I understand this.

When you connect the connection signaling goes through the relay, thus have a ~5s latency for opening the connection.
Once it is open, you have a direct connection that have normal latency.

File: docs/how-to/nat-configuration.md [hunk 7]
```diff
@@ -1,32 +1,170 @@
...
+- Connection success rate is sub-optimal when someone connects to you because different routers models react differently.
+
+For a deeper dive into how holepunching works in Kubo, it's drawbacks, and more information on using DCUtR, see the [blog post](https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/).
+
+Because of these drawbacks, you may want to disable DCUtR holepunching and use another solution, like manual port forwarding. To disable DCUtR holepunching, set [`Swarm.EnableHolePunching`]( https://github.com/ipfs/kubo/blob/master/docs/config.md#swarmenableholepunching) to `false`:
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:52]: Again, you don't need to disable it, it's trying to answer:
> It works (kinda), why would I want something better ?

File: docs/how-to/nat-configuration.md [hunk 8]
```diff
@@ -1,32 +1,170 @@
...
+
+1. Open your Kubo configuration file. 
+
+   :::tip
+   The default location for the config file is `~/.ipfs/config`. If you have set `$IPFS_PATH`, you can find your config file at `$IPFS_PATH/config`.
+   :::
+   
+1. Find the entry for `EnableHolePunching` in `Swarm` or create an entry for `EnableHolePunching` in `Swarm`, and set it to `false`:
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:60]: See message above.

**Review by Jorropo — CHANGES_REQUESTED** [review]
I belive that the last round, looks almost perfect :wink:

**Comment by ElPaisano** [comment]
@Jorropo 

Regarding the DCUtR section issues: I rewrote it to (hopefully) clarify that
- it's enabled by default 
- it has some performance and reliability issues 
- setting up manual port forwarding will "disable" DCUtR and address the perf / reliability issues https://bafybeieha5qk2iwojsrpmtufuz7ke4efrnbba7lr4k7bnxjozab2v7uwja.on.fleek.co/how-to/nat-configuration/#use-dcutr-holepunching


Regarding your note that IPv6 does not always solve everything:
Updated IPv6 section with a note https://bafybeieha5qk2iwojsrpmtufuz7ke4efrnbba7lr4k7bnxjozab2v7uwja.on.fleek.co/how-to/nat-configuration/#enable-ipv6

File: docs/how-to/nat-configuration.md [hunk 9]
```diff
@@ -1,32 +1,132 @@
...
 
-## Port forwarding
 
-Each router has different options and solutions for port forwarding. Most router manufacturers have guides for setting up port forwarding on their devices. In general, the steps are something like this:
+### Use DCUtR Holepunching
+
+As of Kubo v0.13, [DCUtR hole punching is enabled by default](https://github.com/ipfs/kubo/blob/master/docs/changelogs/v0.13.md#-relay-v2-client-with-auto-discovery-swarmrelayclient).
+
+DCUtR holepunching has various drawbacks and tradeoffs. Currently, the connection signaling goes through a relay, which can cause an average latency of 5 seconds when opening a connection. Once a direct connection is established, latency is generally normal. Additionally, DCUtR holepunching does not have a 100% success rate and can fail. For a deeper dive into how holepunching works in Kubo, it's drawbacks, and more information on using DCUtR, see the [_Hole punching in libp2p - Overcoming Firewalls_ blog post](https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/).
```
Comment (Jorropo) [inline docs/how-to/nat-configuration.md:48]:
```suggestion
DCUtR holepunching has various drawbacks and tradeoffs. Currently, the connection signaling goes through a relay, which add an average latency of 5 seconds when opening a connection. Once a direct connection is established, latency is normal. Additionally, DCUtR holepunching does not have a 100% success rate and can fail. For a deeper dive into how holepunching works in Kubo, it's drawbacks, and more information on using DCUtR, see the [_Hole punching in libp2p - Overcoming Firewalls_ blog post](https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/). 
```

**Review by Jorropo — APPROVED** [review]
LGTM (except one slight wording about DCUtR sorry I'm being anoying)

File: docs/how-to/nat-configuration.md [hunk 10]
```diff
@@ -1,32 +1,132 @@
...
+
+As of Kubo v0.13, [DCUtR hole punching is enabled by default](https://github.com/ipfs/kubo/blob/master/docs/changelogs/v0.13.md#-relay-v2-client-with-auto-discovery-swarmrelayclient).
+
+DCUtR holepunching has various drawbacks and tradeoffs. Currently, the connection signaling goes through a relay, which adds an average latency of 5 seconds when opening a connection. Once a direct connection is established, latency is normal. Additionally, DCUtR holepunching does not have a 100% success rate and can fail. For a deeper dive into how holepunching works in Kubo, it's drawbacks, and more information on using DCUtR, see the [_Hole punching in libp2p - Overcoming Firewalls_ blog post](https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/). 
+
+Because of these drawbacks, you may want to use another solution, like manual port forwarding. To enable manual port forwarding, see the instructions below.
+
+### Enable manual port forwarding
+
+If your router does not support UPNP and/or IPv6, or you want better reliability and performance than what DCUtR provides, set up manual port forwarding. Complete the following steps to enable manual port forwarding:
```
Comment (TMoMoreau) [inline docs/how-to/nat-configuration.md:54]: Change "UPNP" to "UPnP", to be consistent with the rest of the doc.

File: docs/how-to/nat-configuration.md [hunk 11]
```diff
@@ -1,32 +1,132 @@
...
 1. Enter the IP address of your IPFS node.
-1. Set traffic to go through port `4001`.
+1. Set traffic to go through an outside port. 
+   :::tip
+   If you are unsure of what port to use, `4001` is recommended.
+   :::
 1. Reboot your IPFS node for the changes to take effect. Make sure to reboot the entire machine, not just the IPFS daemon.
 
-Assuming you're not trying to expose your daemon's API to the public internet, opening port `4001`/`tcp` should be sufficient.
+Now that you've opened a port from the internet to your internal Kubo node, update your Kubo configuration to set `Swarm.AppendAnnounce` as a list of addresses that other IPFS nodes will try to contact you at.
```
Comment (TMoMoreau) [inline docs/how-to/nat-configuration.md:75]: I think this paragraph could likely be removed completely. If I'm not mistaken, it is just an introduction to the next step, but the next paragraph already includes an introduction to the next step.

I think either remove it, or move it so that it follows the next header, and reword it to avoid repeating the same information.

File: docs/how-to/nat-configuration.md [hunk 12]
```diff
@@ -1,32 +1,132 @@
...
+
+#### Update the Kubo Configuration
+
+In this step, you will update your Kubo configuration to set `Swarm.AppendAnnounce` as a list of addresses that other IPFS nodes will try to contact you at. This list lets other nodes on the network that the port forward you created in the previous step know that your node exists, and that these are the addresses at which you can be contacted at. To update your configuration:
```
Comment (TMoMoreau) [inline docs/how-to/nat-configuration.md:79]: Something went wonky with this paragraph.

"In this step, you will update your Kubo configuration to set `Swarm.AppendAnnounce` as a list of addresses that other IPFS nodes will try to contact you at. This list lets other nodes on the network (**know**) that the port forward you created in the previous step ~~know that your node~~ exists, and that these are the addresses at which you can be contacted at.

Something like that.

File: docs/how-to/nat-configuration.md [hunk 13]
```diff
@@ -1,32 +1,132 @@
...
+   ```json
+   "AppendAnnounce": [
+     "/ip4/1.2.3.4/tcp/12345",
+     "/ip4/1.2.3.4/udp/12345/quic",
+     "/ip4/1.2.3.4/udp/12345/quic-v1",
+     "/ip4/1.2.3.4/udp/12345/quic-v1/webtransport",
+    ],
+   ```
+
+Now that you've updated your Kubo configuration, restart Kubo so that your node is reachable, and check that the outside port is open.
```
Comment (TMoMoreau) [inline docs/how-to/nat-configuration.md:127]: I would suggest moving this line to follow the next header, as it is essentially an introduction to the next steps.

File: docs/how-to/nat-configuration.md [hunk 14]
```diff
@@ -1,32 +1,132 @@
...
+
+#### Restart your Kubo node
+
+1. Reboot your IPFS node for the changes to take effect. Make sure to reboot the entire machine, not just the IPFS daemon. Once Kubo restarts, your node should be reachable. 
+1. Check that the outside port is open.
```
Comment (TMoMoreau) [inline docs/how-to/nat-configuration.md:132]: Consider adding an additional step to this, one that describes how the user would check that the outside port is open.

**Review by TMoMoreau — CHANGES_REQUESTED** [review]
Requesting some changes to the style, formatting, etc.

Should be good to go after these are addressed.

File: docs/how-to/nat-configuration.md (hunk 11 above)
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:75]: Removing it make most sense to me

File: docs/how-to/nat-configuration.md (hunk 12 above)
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:79]: Your rewrite makes sense, I vote yes

File: docs/how-to/nat-configuration.md (hunk 13 above)
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:127]: Agreed

File: docs/how-to/nat-configuration.md (hunk 14 above)
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:132]: I agree. However, I don't think this should block adding the rest of this in. I'd vote push as is, and then we can update that step early next week

File: docs/how-to/nat-configuration.md (hunk 10 above)
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:54]: Agreed

File: docs/how-to/nat-configuration.md [hunk 15]
```diff
@@ -1,32 +1,132 @@
...
+
+- If your router supports them, [enable IPv6](#enable-ipv6) or [enable UPnP](#enable-upnp) to solve most connection issues
+- [Use DCUtR holepunching](#use-dcutr-holepunching), which is enabled by default as of Kubo v0.13
+- If IPv6 or UPnP are not available, or DCUtR holepunching doesn't meet your performance and relaibility requirements, [enable manual port forwarding](#enable-manual-port-forwarding)
+
+### Enable IPv6
 
-## IPv6
+:::callout
+While enabling IPv6 can fix many connection issues, enabling it may not address all of your connection issues. For example,  your router might support IPv6, but might still be running a firewall that blocks inbound connections.
```
[the remark this answers is not in the document]
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:34]: Agreed

File: docs/how-to/nat-configuration.md [hunk 16]
```diff
@@ -1,32 +1,132 @@
...
+title: Configure NAT and Port Forwarding
 description: Fix common connection issues by managing your routers NAT and forwarding ports so other IPFS nodes can interact with your node.
 ---
 
-# NAT configuration
+# Configure NAT and Port Forwarding
 
-A common issue for new users running IPFS, especially on home networks, is that their IPFS node is _sitting_ behind a NAT (network address translation). It's common for people who run an IPFS node on their home network to have very long wait times or a high rate of request failure. This is because the rest of the nodes on the IPFS network have difficulty connecting to them through their NAT. The NAT allows many machines to share a single public address, and it is essential for the continued functioning of the IPv4 protocol, which would otherwise be unable to serve the needs of the modern networked population with its 32-bit address space.
+A common issue for new users running IPFS Kubo, especially on home networks, is that their IPFS node is behind a [NAT (network address translation)](../concepts/glossary.md#nat) layer (such as a residential network router) experience very long wait times or high rate of request failure.
```
[the remark this answers is not in the document]
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:8]: I like this rewrite, I vote yes

File: docs/how-to/nat-configuration.md [hunk 17]
```diff
@@ -1,32 +1,132 @@
 ---
-title: NAT configuration
+title: Configure NAT and Port Forwarding
 description: Fix common connection issues by managing your routers NAT and forwarding ports so other IPFS nodes can interact with your node.
 ---
 
-# NAT configuration
+# Configure NAT and Port Forwarding
```
[the remark this answers is not in the document]
Comment (ElPaisano) [reply docs/how-to/nat-configuration.md:6]: Agreed

**Comment by ElPaisano** [comment]
> Requesting some changes to the style, formatting, etc.
> 
> Should be good to go after these are addressed.

@TMoMoreau nice review, I appreciate those catches 🔍 I agree with the majority of your changes. Take a look at individual comments.

**Comment by TMoMoreau** [comment]
+1 for getting this in, and addressing the additional step next week.

**Review by TMoMoreau — APPROVED** [review]
Looks good, merge in when ready.
